### ISSUE TITLE
Test does not resolve esm dependency correctly

### ISSUE THREAD

**Issue by Kimbo0o** [vitest-dev/vitest#4007]
### Describe the bug

I have a project with a npm dependency ("feelin" in node_modules) that has esm and commonjs files. The package.json defines "type": "module", so vitest should use the esm file.
When I start my tests with vitest I get an error: "require() of ES Module ... not supported."
This leads me to the conclusion that vitest uses the commonjs file instead of the ESM file, but it expects an ESM file.

Instead vitest should utilize the ESM file.

### Reproduction

https://stackblitz.com/edit/vitejs-vite-jefmmr?file=package.json

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i7-12800H
    Memory: 10.89 GB / 15.48 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
    npm: 8.11.0 - ~/.nvm/versions/node/v16.16.0/bin/npm
  npmPackages:
    @vitejs/plugin-vue: ^4.2.3 => 4.2.3 
    @vitest/coverage-v8: ^0.33.0 => 0.33.0 
    @vitest/ui: ^0.33.0 => 0.33.0 
    vite: ^4.4.6 => 4.4.6 
    vitest: ^0.33.0 => 0.33.0
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
https://www.npmjs.com/package/feelin?activeTab=code

`feelin` specifies type as `module`, but its entry point uses `require`. This package is bundled incorrectly.

**Comment by Kimbo0o** [comment]
Vite dev server and production build work fine with this dependency. Can I change the vitest config to handle the dependency like vite dev server?

**Comment by sheremet-va** [comment]
> Vite dev server and production build work fine with this dependency. Can I change the vitest config to handle the dependency like vite dev server?

Do you use the latest version? If so, you can add a dependency that uses `feelin` to `test.deps.optimizer.web.include` and it should be optimized like Vite does for the browser.

**Comment by Kimbo0o** [comment]
Thank you for that recommendation and also for commenting my Issue in the feelin repo.

I forked the Stackblitz, here is the new version: https://stackblitz.com/edit/vitejs-vite-3r8j67?file=src%2FApp.vue,vite.config.ts
Newest versions of Vite and Vitest, added the depency that relies von `feelin` to `test.deps.optimizer.web.include`. I also tried to add feelin and feelers to the include as well.
I still get the same error.

**Comment by sheremet-va** [comment]
> Thank you for that recommendation and also for commenting my Issue in the feelin repo.
> 
> I forked the Stackblitz, here is the new version: [stackblitz.com/edit/vitejs-vite-3r8j67?file=src%2FApp.vue,vite.config.ts](https://stackblitz.com/edit/vitejs-vite-3r8j67?file=src%2FApp.vue,vite.config.ts) Newest versions of Vite and Vitest, added the depency that relies von `feelin` to `test.deps.optimizer.web.include`. I also tried to add feelin and feelers to the include as well. I still get the same error.

Looks like optimizer is disabled for you (it's a Vitest bug), but I managed to make it work with this config:

```ts
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
  test: {
    globals: true,
    environment: 'jsdom',
    include: ['./test/**/**.ts'],
    setupFiles: ['./setup.js'],
    deps: {
      optimizer: {
        web: {
          enabled: true,
          include: ['bpmn-js-properties-panel', 'bpmn-js/lib/Modeler'],
        },
      },
    },
  },
});
```

```ts
// setup.js
globalThis.CSS = {
  escape: (str) => str
}
```

**Comment by Kimbo0o** [comment]
It works now with `enabled:  true` - thank you!

**Comment by sheremet-va** [comment]
> To confirm, does `vitest` not respect `module` entrypoints?
> 
> You said that the dependency's entrypoint has a require, but I don't see any `require` in `feelin/dist/index.esm.js`.
> 
> Or, are you saying that `module` is only used if `"type": "module"` is not set, and if it is, then `main` is used?

Yes, by default Vitest doesn't respect `module` field for performance reasons:

- we need to preprocess it to run the code instead of just running code from CJS entry
- valid ESM dependencies are delegated to Node.js which doesn't support "module" field which can create inconsistencies if Vite resolves dep to `module`, but Node.js to `main` (so, nested dependencies will reference different objects)

It's still possible to enable support for `module` field with the config, but it will not be applied for externalized dependencies:

```ts
export default {
  resolve: {
    mainFields: ['module']
  }
}
```

**Comment by IanVS** [comment]
Thanks for the explanation!
